package/src/bin/omicron-package.rs

```diff
@@ -359,23 +382,58 @@ async fn get_package(
         }
         PackageSource::Local { .. } | PackageSource::Composite { .. } => {
             progress.set_message("bundle package".into());
-            package
-                .create_with_progress_for_target(&progress, &target, package_name, &output_directory)
-                .await
-                .with_context(|| {
-                    let msg = format!("failed to create {package_name} in {output_directory:?}");
-                    if let Some(hint) = &package.setup_hint {
-                        format!("{msg}\nHint: {hint}")
-                    } else {
-                        msg
-                    }
-                })?;
+
+            let output_file = match package.output {
+                PackageOutput::Zone { .. } => {
+                    output_directory.join(format!("{}.tar.gz", package_name))
+                }
+                PackageOutput::Tarball => {
+                    package.get_output_path(package_name, output_directory)
+                }
+            };
+
+            let cache = BUILD_CACHE.lock().unwrap();
+            let hash = cache.get(&output_file).map(ToOwned::to_owned);
+            mem::drop(cache);
+
+            if hash.is_none()
+                || !does_hash_match(&output_file, &hash.unwrap()).await
```

Comment:
So I'm theoretically on-board with an approach that uses hashes this way to avoid rebuilding - but don't we need to hash the *inputs* instead of the *outputs*?

For example:

- Suppose we build a **Nexus** package from some input files: (1) `db.rs`, (2) `http.rs`, and (3) `sagas.rs`
- Building this package produces `nexus.tar.gz`, which is then hashed and stored in `build-cache.json`
- If we rebuild without changing anything, the hash of `nexus.tar.gz` matches the hash stored in `build-cache.json`, which is great.
- However, if we update `http.rs` to, say, add a new endpoint...
  - `build-cache.json` will still store the old hash
  - `nexus.tar.gz` still exists, built using the **old** files
  - The hash will still appear to match, right?

Doesn't this basically mean that *all* re-builds where inputs change would be "skipped" incorrectly?

Reply:
So, hashing *inputs* is 10000% the Right Way to do this; but it requires a lot more work for various reasons, and not that I am opposed to doing said work, I was just tired of not shipping *any* wins and wanted to get something going. Incremental progress is better than no progress.

My hope was to get something that would re-build spuriously more often than it should in a world where the inputs are hashed, but still usefully cache some of the time. That said, I think you're 100% right that this doesn't actually do this. Sigh. Time to go back to the drawing board.